### ISSUE TITLE
Malformed test_config.yaml causes a crash with misleading Python traceback

### ISSUE THREAD

**Issue by atilante** [apluslms/python-grader-utils#35]
Graderutils version 3.4, tested with container `apluslms/grade-python:3.7-3.5-3.4`.

Comment out line
`  tasks:`
under `validation` in the following example of file *test_config.yaml*.

```
test_groups:
  - module: tests
    description: Local tests

validation:
  display_name: "Is the file structurally correct?"
  tasks:
  - type: python_import
    display_name: "The file can be imported into a Python interpreter."
    file: eratosthenes.py
  - type: python_import
    display_name: "The file has a function named `sieve_of_eratosthenes`."
    file: eratosthenes.py
    attrs:
      sieve_of_eratosthenes: function
```

Result: Graderutils crashes with the following Python traceback:
```
/usr/local/lib/python3.7/dist-packages/python_jsonschema_objects/__init__.py:53: UserWarning: Schema version http://json-schema.org/draft-07/schema not recognized. Some keywords and features may not be supported.
  self.schema["$schema"]
Traceback (most recent call last):
  File "/usr/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/usr/local/lib/python3.7/dist-packages/graderutils/main.py", line 205, in <module>
    grading_json = run(config_path, **cli_args)
  File "/usr/local/lib/python3.7/dist-packages/graderutils/main.py", line 166, in run
    if "feedback_template" in config:
TypeError: argument of type 'NoneType' is not iterable
Traceback (most recent call last):
  File "/usr/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/usr/local/lib/python3.7/dist-packages/graderutils_format/html.py", line 62, in <module>
    grading_data = json.load(sys.stdin)
  File "/usr/lib/python3.7/json/__init__.py", line 296, in load
    parse_constant=parse_constant, object_pairs_hook=object_pairs_hook, **kw)
  File "/usr/lib/python3.7/json/__init__.py", line 348, in loads
    return _default_decoder.decode(s)
  File "/usr/lib/python3.7/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib/python3.7/json/decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
```

**Context**: I encountered this when updating grading containers for Python programming exercise from `apluslms/grade-python:3.5-2.2` to `apluslms/grade-python:3.7-3.5-3.4`. I needed to modify the existing *test_config.yaml* file for the exercise, and made a human error, forgetting to add the `tasks` line.

**What seems to happen**: the Python JSON Schema library silently fails to parse the test_config.yaml, because it thinks it is malformed, and returns `None`. Only later Graderutils tries to read the configuration in main.py line 166, and fails obviously.

**Proposal**: add an user-friendly error message such as "Python-grader-utils: error in test_config.yaml line X: entry Y is missing. For more information, see the Python-grader-utils manual at https://github.com/apluslms/python-grader-utils/blob/master/graderutils/README.md."

This seem to be related to https://github.com/apluslms/python-grader-utils/issues/17 .

**Comment by Mankro** [comment]
Yeah, grader-utils should validate the structure of test_config.yaml and give clear error messages. @ihalaij1 what do you think?

**Comment by ihalaij1** [comment]
There is validation of the test_config.yaml, but it does not seem to work. This is probably caused by a mismatch between the used test config schema versions on the following lines (in the most recent Graderutils v4.1):
https://github.com/apluslms/python-grader-utils/blob/4845b37b3bae08ee7b473ce9fea6942981bc3e02/graderutils/schemaobjects.py#L17
https://github.com/apluslms/python-grader-utils/blob/4845b37b3bae08ee7b473ce9fea6942981bc3e02/graderutils_format/schemabuilder.py#L35

And a similar problem in Graderutils v3.5:
https://github.com/apluslms/python-grader-utils/blob/4f4265129d5b84a73915b2798c53fdb0630c56e3/graderutils/schemaobjects.py#L12
https://github.com/apluslms/python-grader-utils/blob/4f4265129d5b84a73915b2798c53fdb0630c56e3/graderutils_format/schemabuilder.py#L35

This should be updated so that the default schema version has to be set only in one place, because its difficult to remember every line that needs to be updated when the schema gets changed (as it did recently).

I can make these changes.

**Comment by ihalaij1** [comment]
There is validation test_config.yaml, but it does not seem to work correctly:
https://github.com/apluslms/python-grader-utils/blob/4845b37b3bae08ee7b473ce9fea6942981bc3e02/graderutils/main.py#L127

I will look closer and figure out what's wrong with it.

**Comment by ihalaij1** [comment]
It turns out the test_config.yaml schema validation works just fine, but the issue is that in the example you provided the yaml parser fails instead and `yaml.parser.ParserError` wasn't handled correctly in Graderutils on the following lines:
https://github.com/apluslms/python-grader-utils/blob/4845b37b3bae08ee7b473ce9fea6942981bc3e02/graderutils/main.py#L122-L124
https://github.com/apluslms/python-grader-utils/blob/4845b37b3bae08ee7b473ce9fea6942981bc3e02/graderutils/main.py#L172-L173

I will make a pull request for the fixes.

**Comment by atilante** [comment]
Nice! That was fast, thanks! I'll test this myself at some point, no hurry.

**Comment by ihalaij1** [comment]
> Nice! That was fast, thanks! I'll test this myself at some point, no hurry.

This fix is now released in images `apluslms/grade-python:3.7-4.2-3.5` and `apluslms/grade-python:3.9-4.2-4.0`.
The first one uses grading-base 3.5 and Python 3.7.
The second one uses grading-base 4.0 and Python 3.9.
